include/realizations/catchment/Bmi_Py_Formulation.hpp

```diff
@@ -95,7 +71,7 @@ namespace realization {
 
     protected:
 
-        std::shared_ptr<models::bmi::Bmi_Py_Adapter> construct_model(const geojson::PropertyMap &properties) override;
+        std::shared_ptr<models::bmi::Bmi_Adapter> construct_model(const geojson::PropertyMap &properties) override;
```

Comment:
I'm curious about this change. What's the reason to return the base class here instead of the derived class? Since it's covariant, it should be okay? (though, probably superfluous)

Reply:
The signature has to exactly match the parent class, since this is a `virtual` member function.

Part of the change is to remove the templating of the base class that had this returning the derived concrete class.

Reply:
Ah, I see. I thought covariant return types worked with smart pointers, but sadly they don't i.e. https://godbolt.org/z/nGE9rjEe6. Yeah nevermind then 🙂

Reply:
They don't, but keep in mind that's not why I made the change - as I said, the base class declaring this virtual function doesn't know the concrete type anymore (because I dropped the template), and the signature has to precisely match the base class's declaration of this function.